src/foam/nanos/crunch/ui/CapabilityView.js

```diff
@@ -0,0 +1,172 @@
+        this.wizardletSectionsList = this.wizardletSectionsList.concat(curWizardletSectionsList);
+      }
+    },
+
+    async function initE() {
+      const self = this;
+
+      this.start().addClass(this.myClass())
+        .add(self.slot(function(wizardletSectionsList) {
+          return this.E().forEach(wizardletSectionsList, function(sections, index) {
+            sections.map(section => (
+              this.tag(self.SectionView, {
+                section,
+                data: self.wizardlets[index].data,
+                showTitle: self.showTitle
+              })
+            ));
+          });
+        }))
+      .end();
+    },
+
+    // add a listener to wizardlet.data and each wizardlet.data obj
+    // for the purpose of saving all user inputs releasing
+    // calling views from the responsibility
+    function addListeners(wizardlet, obj, depth=0) {
+      // Some obj (e.g., dao) have a long chain of nested objects
+      // which causes a stack overflow. Such objects don't need listeners
+      // as they are dependent on other objects that are updated by user inputs
+      // so we set depth to prevent a stack overflow while setting listeners on
+      // all the necessary objects
+      if ( ! this.isFObject(obj) || depth > this.MAX_DEPTH ) return;
+
+      // add listeners on wizardlet data
+      obj.sub(this.updateWizardlet.bind(this, wizardlet));
+
+      // add listeners on inner objects for obj
+      for ( let innerObj of Object.values(obj.instance_) ) {
+        if ( this.isFObject(innerObj) ) {
+          this.addListeners(wizardlet, innerObj, ++depth);
+        }
+      }
+    },
+
+    function isFObject(obj) {
+      return typeof obj === 'object' && obj['instance_'] !== undefined;
+    },
+
+    function populateData(wizardlet) {
+      // get all the properties for this wizardlet
+      const properties = wizardlet.capability.of.getAxiomsByClass(foam.core.Property);
+
+      // add property to wizardlet data if the property is an object and does not
+      // exist in wizardlet data
+      for ( let p of properties ) {
+        if ( ! wizardlet.data[p.name] && p.of ) {
+          let pClassName = p.of.id;
+          wizardlet.data[p.name] = foam.lookup(pClassName).create({}, this);
+        }
+      }
+    },
+
+    function saveNoDataCaps(capabilities) {
+      return new Promise(wizardResolve => {
+        // save no-data capabilities (i.e. not displayed in wizard)
+        Promise.all(capabilities.filter(cap => ! cap.of).map(
+          cap => this.userCapabilityJunctionDAO.put(this.UserCapabilityJunction.create({
+            sourceId: this.subject.user.id,
+            targetId: cap.id
+          }))
+        )).then(() => {
+          wizardResolve();
+        });
+      });
+    },
+  ],
+
+  listeners: [
+    {
+      name: 'updateWizardlet',
+      code: function(wizardlet) {
+        try {
+          wizardlet.save();
+          this.saveNoDataCaps(this.capabilities);
```

Comment:
I'd recommend removing the saveNoDataCaps - and using the crunchController.finalOnClose() function here.

I realize finalOnClose isn't perfect with it's console.log and x.closeDialog() - however neither are going to affect you.

Reply:
I've replaced saveNoDataCaps with finalOnClose. Just a heads up, when we use finalOnClose, the modal closes when the user checks the agreement.